server/src/main/java/org/elasticsearch/common/Rounding.java

```diff
@@ -888,6 +826,133 @@ public boolean equals(Object obj) {
         public String toString() {
             return "Rounding[" + interval + " in " + timeZone + "]";
         }
+
+        private long roundKey(long value, long interval) {
+            if (value < 0) {
+                return (value - interval + 1) / interval;
+            } else {
+                return value / interval;
+            }
+        }
+
+        private class FixedRounding implements Prepared {
+            private final LocalTimeOffset offset;
+
+            FixedRounding(LocalTimeOffset offset) {
+                this.offset = offset;
+            }
+
+            @Override
+            public long round(long utcMillis) {
+                return offset.localToUtcInThisOffset(roundKey(offset.utcToLocalTime(utcMillis), interval) * interval);
+            }
+
+            @Override
+            public long nextRoundingValue(long utcMillis) {
+                // TODO this is used in date range's collect so we should optimize it too
```

Comment:
This is pretty much what we used to do. It ain't fast, but we can make it faster later.